tiatoolbox/tools/patchextraction.py

```diff
@@ -0,0 +1,330 @@
+# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+#
+# The Original Code is Copyright (C) 2020, TIALab, University of Warwick
+# All rights reserved.
+# ***** END GPL LICENSE BLOCK *****
+
+"""This file defines patch extraction methods for deep learning models."""
+from abc import ABC
+
+from tiatoolbox.dataloader.wsireader import get_wsireader
+from tiatoolbox.utils.exceptions import MethodNotSupported
+from tiatoolbox.utils.misc import read_point_annotations
+
+# import math
+# import pathlib
+
+
+class PatchExtractor(ABC):
+    """
+    Class for extracting and merging patches in standard and whole-slide images.
+
+    Args:
+        input_img(str, pathlib.Path, ndarray): input image for patch extraction.
+        patch_size(Tuple of int): patch size tuple (width, height).
+        pad_size(Tuple of int): symmetric padding in (x, y) direction.
+
+    Attributes:
+        input_img(ndarray, WSIReader): input image for patch extraction.
+          input_image type is ndarray for an image tile whereas :obj:`WSIReader`
+          for an WSI.
+        patch_size(Tuple of int): patch size tuple (width, height).
+        pad_size(Tuple of int): symmetric padding in (x, y) direction.
+        n(int): current state of the iterator.
+
+    """
+
+    def __init__(self, input_img, patch_size, pad_size=0):
+        self.patch_size = patch_size
+        self.pad_size = pad_size
+        self.n = 0
+        self.wsi = get_wsireader(input_img=input_img)
+
+    def __iter__(self):
+        self.n = 0
+        return self
+
+    def __next__(self):
+        raise NotImplementedError
+
+    def __getitem__(self, item):
+        raise NotImplementedError
+
+    # @staticmethod
+    # def __get_last_steps(image_dim, label_patch_dim, stride):
+    #     """Get the last location for patch extraction in a specific direction.
+    #
+    #     Args:
+    #         image_dim: 1D size of image
+    #         label_patch_dim: 1D size of patches
+    #         stride: 1D size of stride for patch extraction
+    #
+    #     Returns:
+    #         last_step: the final location for patch extraction
+    #     """
+    #     nr_step = math.ceil((image_dim - label_patch_dim) / stride)
+    #     last_step = (nr_step + 1) * stride
+    #     return int(last_step)
+
+    # def extract_patches(self):
+    #     """Extract patches from an image using locations provided by labels data.
+    #
+    #     Args:
+    #         labels (str, ndarray):
+    #
+    #     Returns:
+    #         img_patches (ndarray): extracted image patches of size NxHxWxD.
+    #     """
+    #     raise NotImplementedError
+
+    def merge_patches(self, patches):
+        """Merge the patch-level results to get the overall image-level prediction.
+
+        Args:
+            patches: patch-level predictions
+
+        Returns:
+            image: merged prediction
+        """
+        raise NotImplementedError
+
+
```

Comment:
If merge_patches is a way to accumulate classification results for example, I am not sure if it is the right place for it. I think it is better to keep this class to patch extraction only. Also, based on input arguments, we do not accept any input to merge here.

Reply:
`merge_patches` will be required in window patch extraction. As the method needs to be implemented in the base class thats why it's here.